### PR TITLE
Add generic tabular data output formatter

### PR DESCRIPTION
**Pull request by tsroten**
## Description

This adds a generic tabular data output formatter. Here is an example:

```python
>>> from cli_helpers.tabular_output import TabularOutputFormatter
>>> formatter = TabularOutputFormatter(format_name='plain')
>>> data = [['1', 'Foo'], ['2', 'Bar']]
>>> headers = ['id', 'name']
>>> output = formatter.format_output(data, headers, disable_numparse=False, preserve_whitespace=False)
>>> print(output)
  id  name
   1  Foo
   2  Bar
```

## Checklist

- [x] I've added this contribution to the `CHANGELOG`.
- [x] I've added my name to the `AUTHORS` file (or it's already there).

### CODE REVIEW

**Comment by tsroten** [comment]
@amjith @meeuw @j-bennet Ok, I've ported over the generic formatter from the mycli branch to this pull request.

Here is the rendered documentation for this branch (see the Quickstart and API pages): http://cli-helpers.readthedocs.io/en/feature-tabular_output/index.html

The documentation does not yet cover registering new formatters or using preprocessors. However, I do think it goes deep enough for a first release.

If you have time, please post any thoughts/feedback. Thanks!

**Comment by meeuw** [comment]
I think the summary is a bit too short, I think we should add something about the objectives / purpose of cli_helpers. The documentation about output formatter is great! 👍

**Comment by tsroten** [comment]
Yeah, good point -- thanks for bringing that up. I hadn't focused much on the overall CLI Helper description in the docs or README. I'll try to put that into a different PR so as to keep this one about the formatter as much as possible.

**Comment by tsroten** [comment]
General project documentation added in #2

File: cli_helpers/_compat.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# -*- coding: utf-8 -*-
+"""Python 2/3 compatibility support."""
```
Comment (j-bennet) [inline cli_helpers/_compat.py]: This compat module is for internal utils, right? Why not move it into `__init__.py`, so you would not have to import things from an underscored package name?

File: cli_helpers/tabular_output/output_formatter.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    def get_format_name(self):
+        """Get the current default format.
+
+        :return: The format name.
+        :rtype: str or None
+
+        """
+        return self._format_name
+
+    def supported_formats(self):
```
Comment (j-bennet) [inline cli_helpers/tabular_output/output_formatter.py]: This may look better as a property.

File: cli_helpers/tabular_output/output_formatter.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+        :return: The format names.
+        :rtype: tuple
+
+        """
+        return tuple(self._output_formats.keys())
+
+    @classmethod
+    def register_new_formatter(cls, format_name, handler, preprocessors=(),
+                               kwargs={}):
```
Comment (j-bennet) [inline cli_helpers/tabular_output/output_formatter.py]: Hmm. I'm not a Python guru so perhaps I misunderstand the purpose of `kwargs={}` here. I thought mutable default parameters is not a good thing. Does this serve some special purpose?

File: cli_helpers/tabular_output/tabulate_adapter.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+from cli_helpers.packages import tabulate
```
Comment (j-bennet) [inline cli_helpers/tabular_output/tabulate_adapter.py]: Missing `# -*- coding: utf-8 -*-`.

File: cli_helpers/tabular_output/terminaltables_adapter.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import terminaltables
```
Comment (j-bennet) [inline cli_helpers/tabular_output/terminaltables_adapter.py]: Missing `# -*- coding: utf-8 -*-`.

File: tests/test_utils.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+
+def test_to_string_non_bytes():
+    """Test that to_string() converts non-bytes to a string."""
+    assert utils.to_string(1) == '1'
+    assert utils.to_string(2.33) == '2.33'
+
+
+def test_intlen_with_deimal():
+    """Test that intlen() counts correctly with a decmial place."""
```
Comment (j-bennet) [inline tests/test_utils.py]: Docstring typo: decmial.

File: cli_helpers/tabular_output/output_formatter.py (hunk 3 above)
Comment (tsroten) [inline cli_helpers/tabular_output/output_formatter.py]: Great point 👍 This was added here due to a bug that @meeuw found in the code. But, you're right that it isn't a good idea to use mutable default values. I'll get that fixed.

File: cli_helpers/tabular_output/output_formatter.py (hunk 2 above)
Comment (tsroten) [inline cli_helpers/tabular_output/output_formatter.py]: Good idea!

File: cli_helpers/_compat.py (hunk 1 above)
Comment (tsroten) [inline cli_helpers/_compat.py]: If we move it to `__init__.py`, then our internal utilities would clutter up the public namespace -- e.g. `from cli_helpers import *`.

It seems like it's fairly common practice to use underscored modules internally (e.g. Click, Flask, Werkzeug, Django, and CherryPy).

However, it's also common practice to name internal modules without a leading underscore (e.g. SQLAlchemy and Python Prompt Toolkit).

I'm open to the leading or not-leading underscore. Maybe it's more consistent in dbcli's packages to drop the leading underscore?

File: cli_helpers/tabular_output/tabulate_adapter.py (hunk 4 above)
Comment (tsroten) [inline cli_helpers/tabular_output/tabulate_adapter.py]: fixed

File: cli_helpers/tabular_output/terminaltables_adapter.py (hunk 5 above)
Comment (tsroten) [inline cli_helpers/tabular_output/terminaltables_adapter.py]: fixed

File: cli_helpers/_compat.py (hunk 1 above)
Comment (j-bennet) [inline cli_helpers/_compat.py]: I'd probably drop the underscore.

Comment (tsroten) [inline cli_helpers/_compat.py]: Sounds good, I dropped the underscore. Thanks for looking at this pull request. I'm going to go ahead and merge this in tomorrow as long as no other comments come up.

**Comment by j-bennet** [comment]
👍

**Pull request merged**